# Base Model
Our base model recommends the most popular movies with rating over 3. There is no personalisation based on ther user's history.

In [24]:
import numpy as np
import pandas as pd

In [25]:
df = pd.read_csv('user_train_df.csv')
dft = pd.read_csv('user_test_df.csv')

all_data= pd.concat([df, dft], ignore_index=True)

Find the movies that would be recommended to all users

In [26]:
# Filter movies with ratings greater than 3
filtered_df = df[df["Rating"] > 3]

# Count occurrences of each movie
movie_counts = filtered_df["Movie Title"].value_counts()

# Get the top 10 most common movies
top_10_movies = movie_counts.head(10)

# Display the results
print(top_10_movies)

Return of the Jedi (1983)                 379
Raiders of the Lost Ark (1981)            348
Toy Story (1995)                          321
Empire Strikes Back, The (1980)           293
Twelve Monkeys (1995)                     263
Air Force One (1997)                      252
Princess Bride, The (1987)                248
Jerry Maguire (1996)                      246
Monty Python and the Holy Grail (1974)    242
L.A. Confidential (1997)                  239
Name: Movie Title, dtype: int64


In [37]:
# Calculate the average rating for each movie
average_ratings =df.groupby("Item ID")["Rating"].mean()

# Count occurrences of each movie
movie_counts = df["Item ID"].value_counts()

# Get the top 10 most common movies
top_10_movies = movie_counts.head(10)

# Get the average ratings for the top 10 movies
top_10_movies_with_avg_ratings = average_ratings.loc[top_10_movies.index]

av_ratings = average_ratings.to_frame()

df_ranking = average_ratings.sort_values(ascending=False).reset_index()


(df_ranking.head(10))

Item ID    Rating
0     1653  5.000000
1     1500  5.000000
2     1467  5.000000
3     1599  5.000000
4     1449  4.625000
5     1398  4.500000
6     1642  4.500000
7     1594  4.500000
8      169  4.466102
9       12  4.385768

In [27]:
# Calculate the average rating for each movie
average_ratings =df.groupby("Movie Title")["Rating"].mean()

# Count occurrences of each movie
movie_counts = df["Movie Title"].value_counts()

# Get the top 10 most common movies
top_10_movies = movie_counts.head(10)

# Get the average ratings for the top 10 movies
top_10_movies_with_avg_ratings = average_ratings.loc[top_10_movies.index]

av_ratings = average_ratings.to_frame()

In [28]:
av_ratings

Rating
Movie Title                               
 kldum klaka (Cold Fever) (1994)  3.000000
101 Dalmatians (1996)             2.908257
12 Angry Men (1957)               4.344000
2 Days in the Valley (1996)       3.225806
8 Heads in a Duffel Bag (1997)    3.250000
...                                    ...
Wrong Trousers, The (1993)        4.466102
Wyatt Earp (1994)                 3.100000
You So Crazy (1994)               3.000000
Young Frankenstein (1974)         3.945000
Young Guns (1988)                 3.207921

[601 rows x 1 columns]

Now calculate the average MSE for all users based on the average rating from the training data, and comparing to the real ratings in the test data

In [30]:
# Merge the DataFrame with the average ratings
new_df = dft.merge(average_ratings.rename("avg_rating"), on="Movie Title")

# Calculate squared differences
new_df["squared_error"] = (new_df["Rating"] - new_df["avg_rating"]) ** 2

# Group by User ID and calculate the mean squared error for each user
user_mse = new_df.groupby("User ID")["squared_error"].mean()

# Convert the results to a list
mse_list = user_mse.tolist()

#print("MSE for each user:", mse_list)

user_mse.mean()

1.073672076930744

We have an MSE of 1.07. We look to improve upon this with our recommender models